pkg/openwsn/contrib/radio_hal.c

```diff
@@ -0,0 +1,316 @@
+{
+    (void)tx_or_rx;
+
+    ieee802154_phy_conf_t conf =
+    { .channel = frequency,
+      .page = CONFIG_IEEE802154_DEFAULT_CHANNEL,
+      .pow = CONFIG_IEEE802154_DEFAULT_TXPOWER };
+
+    ieee802154_radio_config_phy(openwsn_radio.dev, &conf);
+}
+
+void radio_rfOn(void)
+{
+    ieee802154_radio_request_on(openwsn_radio.dev);
+    ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
+                                           IEEE802154_TRX_STATE_TRX_OFF);
+    /* If the radio is still not in TRX_OFF state, spin */
+    while (ieee802154_radio_confirm_on(openwsn_radio.dev) == -EAGAIN) {}
+}
+
+void radio_rfOff(void)
+{
+    /* radio_rfOff is called in the middle of a slot and is not always
+       followed by an `radio_rfOn`, so don't call `ieee802154_radio_off`
+       and only send the radio to `TrxOFF` instead */
+    int ret = ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
+                                                     IEEE802154_TRX_STATE_TRX_OFF);
+
+    if (ret) {
+        LOG_ERROR("[openwsn/radio]: request_set_trx_state failed %s\n",
+                  __FUNCTION__);
+    }
+    else {
+        debugpins_radio_clr();
+        leds_radio_off();
+        while (ieee802154_radio_confirm_set_trx_state(openwsn_radio.dev) ==
+               -EAGAIN) {}
+    }
+
+}
+
+void radio_loadPacket(uint8_t *packet, uint16_t len)
+{
+    /* OpenWSN `len` accounts for the FCS field which is set by default by
+       netdev, so remove from the actual packet `len` */
+    iolist_t pkt = {
+        .iol_base = (void *)packet,
+        .iol_len = (size_t)(len - IEEE802154_FCS_LEN),
+    };
+
+    if (ieee802154_radio_write(openwsn_radio.dev, &pkt) < 0) {
+        LOG_ERROR("[openwsn/radio]: couldn't load pkt\n");
+    }
+}
+
+void radio_txEnable(void)
+{
+    int ret = ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
+                                                     IEEE802154_TRX_STATE_TX_ON);
+
+    if (ret) {
+        LOG_ERROR("[openwsn/radio]: request_set_trx_state failed %s\n",
+                  __FUNCTION__);
+    }
+    else {
+        while (ieee802154_radio_confirm_set_trx_state(openwsn_radio.dev) ==
+               -EAGAIN) {}
+        debugpins_radio_set();
+        leds_radio_on();
+    }
+}
+
+void radio_txNow(void)
+{
+    int ret = ieee802154_radio_request_transmit(openwsn_radio.dev);
+
+    if (ret) {
+        LOG_ERROR("[openwsn/radio]: request_set_trx_state failed %s\n",
+                  __FUNCTION__);
+    }
+    else {
+        /* Trigger startFrame manually if no IEEE802154_CAP_IRQ_TX_START */
+        if (!ieee802154_radio_has_irq_tx_start(openwsn_radio.dev)) {
+            debugpins_isr_set();
+            _txrx_event_capture_time = sctimer_readCounter();
+            debugpins_isr_clr();
+            openwsn_radio.startFrame_cb(_txrx_event_capture_time);
+        }
+    }
+}
+
+void radio_rxEnable(void)
+{
+    int ret = ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
```

Comment:
tip here: Usually going from `TX_ON` to `RX_ON` is way faster than `TRX_OFF` to `RX_ON` (sometimes a couple of microsecs). Maybe you could set the state to TX_ON here

Reply:
alternatively, I was thinking that adding a `IEEE802154_RX_DISABLED` to `set_rx_mode` could help to disable the detection of SFD while the RX circuit is still on. What do you think?

Reply:
@jia200x I would prefer the second approach since its explicit, but for now I can go with your first suggestion!

Reply:
done